proposals/heps/0002-YAML-converter.md

```diff
@@ -0,0 +1,471 @@
+            )
+```
+
+### Scripts
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: scripts-python-
+spec:
+  entrypoint: python-script-example
+  templates:
+  - name: python-script-example
+    steps:
+    - - name: generate
+        template: gen-random-int
+    - - name: print
+        template: print-message
+        arguments:
+          parameters:
+          - name: message
+            value: "{{steps.generate.outputs.result}}"
+
+  - name: gen-random-int
+    script:
+      image: python:alpine3.6
+      command: [python]
+      source: |
+        import random
+        i = random.randint(1, 100)
+        print(i)
+
+  - name: print-message
+    inputs:
+      parameters:
+      - name: message
+    container:
+      image: alpine:latest
+      command: [sh, -c]
+      args: ["echo result was: {{inputs.parameters.message}}"]
+```
+
+```py
+from hera.workflows import Container, Parameter, Script, Step, Steps, Workflow
+
+with Workflow(
+    generate_name="scripts-python-",
+    entrypoint="python-script-example",
+) as w:
+    gen_random_int = Script(
+        name="gen-random-int",
+        image="python:alpine3.6",
+        command=["python"],
+        source="""
+import random
+i = random.randint(1, 100)
+print(i)""",
+    )
+
+    print_message = Container(
+        name="print-message",
+        image="alpine:latest",
+        command=["sh", "-c"],
+        args=["echo result was: {{inputs.parameters.message}}"],
+        inputs=[Parameter(name="message")],
+    )
+
+    with Steps(name="python-script-example") as s:
+        Step(
+            name="generate",
+            template="gen-random-int",
+        )
+        Step(
+            name="print",
+            template="print-message",
+            arguments={"message": "{{steps.generate.outputs.result}}"},
+        )
+```
+
+
+# Implementation
+
+We will give a proof-of-concept for generating the Python object from YAMLs and describe the algorithm for the code:
+
+* Given a Workflow YAML, we know the values contained are limited to basic data types, lists and dictionaries.
+* We should then be able to iterate through the _attributes of the `Workflow` class_ and intelligently map the given YAML values to produce Python assignments, e.g.
+  * Basic data types such as `str` and `int` are assigned directly, e.g. for a Workflow's `name`, we will produce the Python assignment in code as `name="some-name",`
+  * Lists must be iterated through, and by using type of the values in the list, we recursively extract the correct assignment, e.g. a `list[str]` will produce the Python assignment `things=["a", "b", "c"]`, whereas `list[list[str]]` produces `things=[["a", "b"], ["c", "d"]]`
+  * Dictionaries must be iterated through, and by using the type of the values (the keys are always strings coming from YAML), we can recursively extract the correct assignment, e.g. a `dict[str, int]` will produce the assignment `things={"a": 1, "b": 2}`
```

Comment:
We can use `repr` to easily solve all these cases, including correctly escaping the strings.

Reply:
Thanks, couldn't remember how to get them to print nicely as actual Python code